### ISSUE TITLE
Incompatibility Between Assimp 3.3.1 and MATLAB

### ISSUE THREAD

**Issue by tlian7** [benjamin-heasly/mexximp#2]
After updating to Assimp 3.3.1 today we discovered an incompatibility between this new version of Assimp and MATLAB (both R2014b and R2015b). Specifically, when one tries to run makeMexximp.cpp all the tests will fail with an error that looks something like this:

 `Invalid MEX-file '/home/tlian/GitRepos/mexximp/build/mexximpImport.mexa64': /share/software/MATLAB-R2015b/bin/glnxa64/../../sys/os/glnxa64/libstdc++.so.6: version 'GLIBCXX_3.4.19' not found (required by /usr/local/lib/libassimp.so.3)`

It seems like other people have had this error before too, and the reasons and solutions behind it are well documented in [**this Github issue**.](https://github.com/mlapin/libsdca/issues/1)

In a nutshell, it seems that MATLAB loads its own version of _libstdc++_ that is different from the system version. This version is outdated and may not contain more recent symbols (`GLIBCXX_3.4.19') required by the C++ code.

The solution that worked for us was to include a "LD_PRELOAD" before running MATLAB. We wrote this into our bash profile. The solution is detailed below, from the link above:

> Run locate libstdc++.so.6 to find the library on your system.
> Use LD_PRELOAD=/located-path/libstdc++.so.6 matlab to run Matlab.

I wonder if there is a solution that is less temporary than this? In any case, maybe this should be documented somewhere?

**Comment by benjamin-heasly** [comment]
Yes.  Thanks for summarizing this and linking to the post that explains the issue so well.

I think this is a general problem with Matlab and mex-functions that rely on shared libraries.  It's annoying!  I'll add a note to the mexximp README about the workaround with LD_PRELOAD.

Long term, my goal is to stop using mex functions to talk to shared libraries like Assimp.  Instead, we can write stand-alone executables that talk to Matlab over standard IO or file IO.  The stand-alone executables would could "not care" about the Matlab's own versions of libraries like libstdc++, allowing this issue to go away.

It would also be possible to put the stand-alone execuables into Docker containers along with their library dependencies, which would be an added bonus.

This approach would add some communication overhead.  But for projects like mexximp, I suspect most of the time is spent doing file IO and processing of the 3D scene, so the communication overhead might be OK.

**Comment by benjamin-heasly** [comment]
Hi @tlian7 -- 

I added [some documentation of the issue](https://github.com/RenderToolbox/mexximp#note-on-linux-library-version-hell) in the README.

I also made a new issue #7 with a view to preventing this kind of problem in the future.

I'm going to go ahead and close this issue.
